### PR TITLE
Add Bipartite GraphSage example

### PR DESCRIPTION
**Pull request by husimplicity**
This PR adds an example implementation of a recommendation model bipartite Graphsage.
The current implementation is a supervised version on MovieLens. The i2i co-occurrence matrix is applied to construct i2i-graph. When applying Graphsage, the model follows i-i-i pattern to encode item embedding and i-i-u for user-side respectively. Since the task to predict u2i rates is supervised,  we apply MSE loss with labels.
For next step, we will apply this model to an unsupervised task on another dataset with negative sampling and triplet loss.

### CODE REVIEW

File: examples/hetero/bipartite_sage.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+
+path = osp.join(osp.dirname(osp.realpath(__file__)), '../../data/MovieLens')
+user_sim_dict_path = 'swing_user_sim_dict.json'
+dataset = MovieLens(path, model_name='all-MiniLM-L6-v2')
+data = dataset[0].to(device)
+
+data['user'].x = torch.LongTensor(torch.arange(
+    0, data['user'].num_nodes)).to(device)
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:26]:
```suggestion
data = dataset[0]
data['user'].x = torch.arange(data['user'].num_nodes)
data = data.to(device)
```

File: examples/hetero/bipartite_sage.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+train_data, val_data, test_data = T.RandomLinkSplit(
+    num_val=0.1,
+    num_test=0.1,
+    neg_sampling_ratio=0.0,
+    edge_types=[('user', 'rates', 'movie')],
+    rev_edge_types=[('movie', 'rev_rates', 'user')],
+)(data)
+
+
+def to_u2i_mat(edge_index, i_num, u_num):
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:43]: You can actually use `utils.to_dense_adj` for this.

File: examples/hetero/bipartite_sage.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    res_mat = torch.zeros((n_items, num_neighbors))
+    res_sim_mat = torch.zeros((n_items, num_neighbors))
+
+    items_D = torch.sum(A, axis=0).reshape(-1)
+    users_D = torch.sum(A, axis=1).reshape(-1)
+    beta_uD = (1 / torch.sqrt(users_D + 1)).reshape(-1, 1)
+    beta_iD = (1 / torch.sqrt(items_D + 1)).reshape(1, -1)
+    all_ii_constraint_mat = beta_uD @ beta_iD
+
+    for i in range(n_items):
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:65]: Is there a need for this for-loop here? I assume you can apply `topk` in `dim=1` for this as well.

File: examples/hetero/bipartite_sage.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+        row = all_ii_constraint_mat[i] * A[i]
+        row_sims, row_idxs = torch.topk(row, num_neighbors)
+        res_mat[i] = row_idxs
+        res_sim_mat[i] = row_sims
+
+    return res_mat.long(), res_sim_mat.float()
+
+
+def get_i2i_sim_graph(i2imat, i2isimmat):
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:74]: `nonzero()` + masking should work as well, right?

File: examples/hetero/bipartite_sage.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    return torch.tensor(i2i), torch.stack(i2i_sim)
+
+
+u2i_mat = to_u2i_mat(train_data.edge_index_dict[('user', 'rates', 'movie')],
+                     train_data['movie'].x.shape[0],
+                     train_data['user'].x.shape[0])
+i2i_mat, i2i_sim_mat = get_coocur_mat(u2i_mat, 9)
+i2i, _ = get_i2i_sim_graph(i2i_mat, i2i_sim_mat)
+
+# Add the generated i2i graph for high-order information
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:94]: I assume you should use the `edge_index` from `train_data`, `val_data` and `test_data` for this. Currently, you are using it only from `train_data`.

File: examples/hetero/bipartite_sage.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        self.conv1 = SAGEConv(-1, hidden_channels)
+        self.conv2 = SAGEConv(-1, hidden_channels)
+        self.lin1 = Linear(hidden_channels, hidden_channels)
+        self.lin2 = Linear(hidden_channels, out_channels)
+
+    def reset_parameters(self):
+        self.conv1.reset_parameters()
+        self.conv2.reset_parameters()
+        self.lin1.reset_parameters()
+        self.lin2.reset_parameters()
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:126]: Can be dropped IMO.

File: examples/hetero/bipartite_sage.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+    def forward(self, x_dict, edge_index_dict):
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:128]:
```suggestion
    def forward(self, x, edge_index):
```
No need to work with dicts here.

File: examples/hetero/bipartite_sage.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        super().__init__()
+
+        self.conv1 = SAGEConv((-1, -1), hidden_channels)
+        self.conv2 = SAGEConv((-1, -1), hidden_channels)
+        self.conv3 = SAGEConv((-1, -1), hidden_channels)
+        self.lin1 = Linear(hidden_channels, hidden_channels)
+        self.lin2 = Linear(hidden_channels, hidden_channels)
+        self.lin3 = Linear(hidden_channels, out_channels)
+
+    def reset_parameters(self):
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:148]: Can be dropped.

File: examples/hetero/bipartite_sage.py [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+class Model(torch.nn.Module):
+    def __init__(self, input_size, hidden_channels, out_channels):
+        super().__init__()
+        self.item_encoder = ItemGNNEncoder(hidden_channels, out_channels)
+        self.user_encoder = UserGNNEncoder(hidden_channels, out_channels)
+        self.decoder = EdgeDecoder(hidden_channels)
+        self.user_emb = Embedding(input_size, hidden_channels, device=device)
+
+    def reset_parameters(self):
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:200]: Can be dropped.

File: examples/hetero/bipartite_sage.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+        return z.view(-1)
+
+
+class Model(torch.nn.Module):
+    def __init__(self, input_size, hidden_channels, out_channels):
+        super().__init__()
+        self.item_encoder = ItemGNNEncoder(hidden_channels, out_channels)
+        self.user_encoder = UserGNNEncoder(hidden_channels, out_channels)
+        self.decoder = EdgeDecoder(hidden_channels)
+        self.user_emb = Embedding(input_size, hidden_channels, device=device)
```
Comment (rusty1s) [inline examples/hetero/bipartite_sage.py:198]: Move to the top.

**Review by rusty1s — COMMENTED** [review]
This is cool, thanks for adding! Left a few comments.

File: examples/hetero/bipartite_sage.py (hunk 1 above)
Comment (husimplicity) [reply examples/hetero/bipartite_sage.py]: Fixed.

File: examples/hetero/bipartite_sage.py (hunk 2 above)
Comment (husimplicity) [reply examples/hetero/bipartite_sage.py:42]: Sure, good advice!

File: examples/hetero/bipartite_sage.py (hunk 3 above)
Comment (husimplicity) [reply examples/hetero/bipartite_sage.py]: Good suggestion, thx.

File: examples/hetero/bipartite_sage.py (hunk 5 above)
Comment (husimplicity) [reply examples/hetero/bipartite_sage.py:88]: Our concern here is the information leak from val_data and test_data to train_data if we use the whole data to construct graph.

File: examples/hetero/bipartite_sage.py (hunk 6 above)
Comment (husimplicity) [reply examples/hetero/bipartite_sage.py]: dropped.

File: examples/hetero/bipartite_sage.py (hunk 7 above)
Comment (husimplicity) [reply examples/hetero/bipartite_sage.py]: fixed.

File: examples/hetero/bipartite_sage.py (hunk 8 above)
Comment (husimplicity) [reply examples/hetero/bipartite_sage.py]: dropped.

File: examples/hetero/bipartite_sage.py (hunk 9 above)
Comment (husimplicity) [reply examples/hetero/bipartite_sage.py]: dropped.

File: examples/hetero/bipartite_sage.py (hunk 10 above)
Comment (husimplicity) [reply examples/hetero/bipartite_sage.py]: moved.
